// jsgantt-improved/dist/test/unit/helpers.js
"use strict";
/**
 * Shared test helpers for unit tests.
 *
 * Import this module BEFORE any src imports so the DOM shim is installed
 * before TypeScript module initialisation runs.
 *
 *   import '../helpers';           // must be first
 *   import { TaskItem } from '../../src/task';
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeTask = void 0;
// ─── Minimal DOM shim ────────────────────────────────────────────────────────
// TaskItem uses document.createTextNode / createElement to sanitise values.
// In a real browser this strips HTML; here we provide the minimum surface area
// needed to construct a TaskItem without a real browser.
function makeTextNode(val) {
    return {
        data: String(val == null ? '' : val),
        nodeName: '#text',
        childNodes: [],
        hasChildNodes: function () { return false; },
    };
}
function makeElement(tag) {
    var el = {
        nodeName: tag.toUpperCase(),
        className: '',
        innerHTML: '',
        childNodes: [],
        hasChildNodes: function () { return this.childNodes.length > 0; },
        replaceChild: function (newNode, oldNode) {
            var idx = this.childNodes.indexOf(oldNode);
            if (idx !== -1)
                this.childNodes[idx] = newNode;
        },
    };
    return el;
}
global.document = {
    createTextNode: makeTextNode,
    createElement: makeElement,
};
// ─── TaskItem factory ─────────────────────────────────────────────────────────
// Pass Date objects so the constructor takes the `instanceof Date` fast-path,
// bypassing parseDateStr / document.createTextNode on the date strings.
var task_1 = require("../../src/task");
var general_utils_1 = require("../../src/utils/general_utils");
var mockGantt = {
    getDateInputFormat: function () { return 'yyyy-mm-dd'; },
    hashString: general_utils_1.hashString,
};
function makeTask(opts) {
    return new task_1.TaskItem(opts.id, opts.name || ('task-' + opts.id), opts.start || '', // '' → vStart stays null → hasStart() === false
    opts.end || '', 'gtaskblue', '', // pLink
    0, // pMile
    '', // pRes
    0, // pComp
    opts.group || 0, opts.parent, 1, // pOpen
    '', // pDepend
    '', // pCaption
    '', // pNotes
    mockGantt, 0, // pCost
    opts.planStart || null, opts.planEnd || null);
}
exports.makeTask = makeTask;
