// jsgantt-improved/dist/test/unit/task-item.spec.js
"use strict";
/**
 * Unit tests for TaskItem accessor methods.
 */
Object.defineProperty(exports, "__esModule", { value: true });
require("./helpers"); // DOM shim — must be first
var chai_1 = require("chai");
var helpers_1 = require("./helpers");
describe('TaskItem.hasStart / hasEnd', function () {
    it('returns false when no date is set', function () {
        var task = (0, helpers_1.makeTask)({ id: 1, parent: 0 });
        (0, chai_1.expect)(task.hasStart()).to.equal(false);
        (0, chai_1.expect)(task.hasEnd()).to.equal(false);
    });
    it('returns true when an explicit start/end date is provided', function () {
        var task = (0, helpers_1.makeTask)({
            id: 2,
            parent: 0,
            start: new Date('2027-01-01'),
            end: new Date('2027-01-31'),
        });
        (0, chai_1.expect)(task.hasStart()).to.equal(true);
        (0, chai_1.expect)(task.hasEnd()).to.equal(true);
    });
    it('returns true when only plan dates are provided', function () {
        var task = (0, helpers_1.makeTask)({
            id: 3,
            parent: 0,
            planStart: new Date('2027-01-01'),
            planEnd: new Date('2027-01-31'),
        });
        (0, chai_1.expect)(task.hasStart()).to.equal(true);
        (0, chai_1.expect)(task.hasEnd()).to.equal(true);
    });
    it('getStart() on a dateless task still returns a Date (backwards compat)', function () {
        var task = (0, helpers_1.makeTask)({ id: 4, parent: 0 });
        (0, chai_1.expect)(task.getStart()).to.be.instanceOf(Date);
        (0, chai_1.expect)(task.getEnd()).to.be.instanceOf(Date);
    });
    it('getStart() returns the explicit start date when set', function () {
        var start = new Date('2027-06-15');
        var task = (0, helpers_1.makeTask)({ id: 5, parent: 0, start: start, end: new Date('2027-06-30') });
        (0, chai_1.expect)(task.getStart().getTime()).to.equal(start.getTime());
    });
});
