// jsgantt-improved/dist-e2e/issue375-column-visibility.spec.js
"use strict";
/**
 * E2E tests for issue #375: left data table columns not visible when set via setOptions().
 *
 * The issue reports that columns like vShowRes, vShowCost, vShowComp, vShowStartDate,
 * vShowEndDate are not visible even when explicitly enabled via setOptions().
 */
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
const test_1 = require("@playwright/test");
const BASE_URL = (_a = process.env.JSGANTT_E2E_BASE_URL) !== null && _a !== void 0 ? _a : 'http://localhost:8080';
async function gotoDemo(page) {
    const url = BASE_URL.replace(/\/$/, '') + '/demo.html';
    await page.goto(url, { waitUntil: 'load' });
    await page.waitForFunction(() => typeof window.JSGantt !== 'undefined', { timeout: 15000 });
}
function setupGantt(page, options) {
    return page.evaluate((opts) => {
        const container = document.getElementById('embedded-Gantt');
        container.innerHTML = '';
        const g = new window.JSGantt.GanttChart(container, 'week');
        g.setOptions(Object.assign({ vFormat: 'week', vShowDeps: 0, vLang: 'en' }, opts));
        const T = window.JSGantt.TaskItem;
        g.AddTaskItem(new T(1, 'Group A', '2024-01-01', '2024-03-31', 'ggroupblack', '', 0, 'Mario', 0, 1, 0, 1, '', '', '', g));
        g.AddTaskItem(new T(2, 'Task Alpha', '2024-01-01', '2024-01-14', 'gtaskblue', '', 100, 'Alice', 50, 0, 1, 1, '', '', '', g));
        g.AddTaskItem(new T(3, 'Task Beta', '2024-01-15', '2024-02-28', 'gtaskgreen', '', 200, 'Bob', 30, 0, 1, 1, '', '', '', g));
        g.Draw();
        return {
            resHeadingCount: document.querySelectorAll('td.gtaskheading.gresource').length,
            durHeadingCount: document.querySelectorAll('td.gtaskheading.gduration').length,
            compHeadingCount: document.querySelectorAll('td.gtaskheading.gpccomplete').length,
            startHeadingCount: document.querySelectorAll('td.gtaskheading.gstartdate').length,
            endHeadingCount: document.querySelectorAll('td.gtaskheading.genddate').length,
            costHeadingCount: document.querySelectorAll('td.gtaskheading.gcost').length,
        };
    }, options);
}
test_1.test.describe('Issue #375 — column visibility via setOptions()', () => {
    (0, test_1.test)('all columns shown by default (no setOptions overrides)', async ({ page }) => {
        await gotoDemo(page);
        const counts = await setupGantt(page, {});
        (0, test_1.expect)(counts.resHeadingCount, 'resource column heading should be visible by default').toBeGreaterThan(0);
        (0, test_1.expect)(counts.durHeadingCount, 'duration column heading should be visible by default').toBeGreaterThan(0);
        (0, test_1.expect)(counts.compHeadingCount, 'completion column heading should be visible by default').toBeGreaterThan(0);
        (0, test_1.expect)(counts.startHeadingCount, 'start date column heading should be visible by default').toBeGreaterThan(0);
        (0, test_1.expect)(counts.endHeadingCount, 'end date column heading should be visible by default').toBeGreaterThan(0);
        // cost is 0 by default
        (0, test_1.expect)(counts.costHeadingCount, 'cost column heading should be hidden by default').toBe(0);
    });
    (0, test_1.test)('vShowCost=1 via setOptions() makes cost column appear', async ({ page }) => {
        await gotoDemo(page);
        const counts = await setupGantt(page, { vShowCost: 1 });
        (0, test_1.expect)(counts.costHeadingCount, 'cost column heading should be visible when vShowCost=1').toBeGreaterThan(0);
    });
    (0, test_1.test)('vShowRes=0 via setOptions() hides resource column', async ({ page }) => {
        await gotoDemo(page);
        const counts = await setupGantt(page, { vShowRes: 0 });
        (0, test_1.expect)(counts.resHeadingCount, 'resource column heading should be hidden when vShowRes=0').toBe(0);
    });
    (0, test_1.test)('vShowComp=0 via setOptions() hides completion column', async ({ page }) => {
        await gotoDemo(page);
        const counts = await setupGantt(page, { vShowComp: 0 });
        (0, test_1.expect)(counts.compHeadingCount, 'completion column heading should be hidden when vShowComp=0').toBe(0);
    });
    (0, test_1.test)('vShowStartDate=0 and vShowEndDate=0 via setOptions() hides date columns', async ({ page }) => {
        await gotoDemo(page);
        const counts = await setupGantt(page, { vShowStartDate: 0, vShowEndDate: 0 });
        (0, test_1.expect)(counts.startHeadingCount, 'start date column should be hidden when vShowStartDate=0').toBe(0);
        (0, test_1.expect)(counts.endHeadingCount, 'end date column should be hidden when vShowEndDate=0').toBe(0);
    });
    (0, test_1.test)('all optional columns toggled on via setOptions() renders all heading cells', async ({ page }) => {
        await gotoDemo(page);
        const counts = await setupGantt(page, {
            vShowRes: 1,
            vShowDur: 1,
            vShowComp: 1,
            vShowStartDate: 1,
            vShowEndDate: 1,
            vShowCost: 1,
        });
        (0, test_1.expect)(counts.resHeadingCount, 'resource heading visible').toBeGreaterThan(0);
        (0, test_1.expect)(counts.durHeadingCount, 'duration heading visible').toBeGreaterThan(0);
        (0, test_1.expect)(counts.compHeadingCount, 'completion heading visible').toBeGreaterThan(0);
        (0, test_1.expect)(counts.startHeadingCount, 'start date heading visible').toBeGreaterThan(0);
        (0, test_1.expect)(counts.endHeadingCount, 'end date heading visible').toBeGreaterThan(0);
        (0, test_1.expect)(counts.costHeadingCount, 'cost heading visible').toBeGreaterThan(0);
    });
    (0, test_1.test)('all optional columns toggled off via setOptions() hides all non-name headings', async ({ page }) => {
        await gotoDemo(page);
        const counts = await setupGantt(page, {
            vShowRes: 0,
            vShowDur: 0,
            vShowComp: 0,
            vShowStartDate: 0,
            vShowEndDate: 0,
            vShowCost: 0,
        });
        (0, test_1.expect)(counts.resHeadingCount, 'resource heading hidden').toBe(0);
        (0, test_1.expect)(counts.durHeadingCount, 'duration heading hidden').toBe(0);
        (0, test_1.expect)(counts.compHeadingCount, 'completion heading hidden').toBe(0);
        (0, test_1.expect)(counts.startHeadingCount, 'start date heading hidden').toBe(0);
        (0, test_1.expect)(counts.endHeadingCount, 'end date heading hidden').toBe(0);
        (0, test_1.expect)(counts.costHeadingCount, 'cost heading hidden').toBe(0);
    });
    (0, test_1.test)('column data cells respect vShowRes and vShowCost settings', async ({ page }) => {
        await gotoDemo(page);
        const result = await page.evaluate(() => {
            const container = document.getElementById('embedded-Gantt');
            container.innerHTML = '';
            const g = new window.JSGantt.GanttChart(container, 'week');
            g.setOptions({
                vFormat: 'week',
                vLang: 'en',
                vShowRes: 0,
                vShowCost: 1,
            });
            const T = window.JSGantt.TaskItem;
            g.AddTaskItem(new T(1, 'Task A', '2024-01-01', '2024-01-31', 'gtaskblue', '', 500, 'Alice', 80, 0, 0, 1, '', '', '', g));
            g.Draw();
            return {
                resCells: document.querySelectorAll('td.gresource').length,
                costCells: document.querySelectorAll('td.gcost').length,
            };
        });
        (0, test_1.expect)(result.resCells, 'resource cells should be absent when vShowRes=0').toBe(0);
        (0, test_1.expect)(result.costCells, 'cost cells should be present when vShowCost=1').toBeGreaterThan(0);
    });
});
