// jsgantt-improved/dist/test/e2e/gantt.page.js
"use strict";
/**
 * Page Object Model for the jsGantt demo page.
 * Wraps common selectors and helpers used across e2e specs.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GanttPage = exports.DEMO_URL = void 0;
exports.DEMO_URL = 'demo.html';
var GanttPage = /** @class */ (function () {
    function GanttPage(page) {
        this.page = page;
    }
    GanttPage.prototype.goto = function (url) {
        if (url === void 0) { url = exports.DEMO_URL; }
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.page.goto(url, { waitUntil: 'load' })];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, this.page.waitForSelector('#embedded-Gantt .gcharttable', { timeout: 10000 })];
                    case 2:
                        _a.sent();
                        // Allow layout to settle
                        return [4 /*yield*/, this.page.waitForTimeout(400)];
                    case 3:
                        // Allow layout to settle
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Apply CSS zoom to the Gantt container, simulating browser-level zoom.
     * zoom=1.0 restores 100%.
     */
    GanttPage.prototype.applyZoom = function (zoom) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.page.evaluate(function (z) {
                            var el = document.querySelector('#embedded-Gantt');
                            if (el)
                                el.style.zoom = z === 1 ? '' : String(z);
                        }, zoom)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Measure bar positions relative to the chart table's left edge.
     * Returns up to `limit` bars.
     */
    GanttPage.prototype.getBarMetrics = function (limit) {
        if (limit === void 0) { limit = 8; }
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.page.evaluate(function (lim) {
                        var chartTable = document.querySelector('.gcharttable');
                        if (!chartTable)
                            return [];
                        var chartRect = chartTable.getBoundingClientRect();
                        return Array.from(document.querySelectorAll('[id*="bardiv_"]'))
                            .slice(0, lim)
                            .map(function (bar) {
                            var el = bar;
                            var rect = el.getBoundingClientRect();
                            return {
                                id: el.id.replace(/.*bardiv_/, 'bar_'),
                                styleLeft: parseInt(el.style.left || '0', 10),
                                styleWidth: parseInt(el.style.width || '0', 10),
                                viewportLeft: Math.round((rect.left - chartRect.left) * 100) / 100,
                                viewportWidth: Math.round(rect.width * 100) / 100,
                            };
                        });
                    }, limit)];
            });
        });
    };
    /**
     * Measure the last header row's column widths in viewport pixels.
     * The last row is the finest date granularity (day/week/month).
     */
    GanttPage.prototype.getColumnMetrics = function (limit) {
        if (limit === void 0) { limit = 20; }
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.page.evaluate(function (lim) {
                        var headerTable = document.querySelector('.gcharttableh');
                        if (!headerTable)
                            return [];
                        var rows = headerTable.querySelectorAll('tr');
                        var lastRow = rows[rows.length - 1];
                        if (!lastRow)
                            return [];
                        return Array.from(lastRow.querySelectorAll('td'))
                            .slice(0, lim)
                            .map(function (td, i) { return ({
                            index: i,
                            text: td.innerText.trim().slice(0, 10),
                            viewportWidth: Math.round(td.getBoundingClientRect().width * 100) / 100,
                        }); });
                    }, limit)];
            });
        });
    };
    return GanttPage;
}());
exports.GanttPage = GanttPage;
