// jsgantt-improved/dist/test/e2e/issue349-taskname-click-event.spec.js
"use strict";
/**
 * E2E tests for issue #349: Click event on task name via vEvents.taskname.
 *
 * The issue reports that users want to handle clicks on task name cells.
 * The vEvents.taskname callback should fire with (task, event, cell) args
 * when a task name cell is clicked.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
var test_1 = require("@playwright/test");
var BASE_URL = (_a = process.env.JSGANTT_E2E_BASE_URL) !== null && _a !== void 0 ? _a : 'http://localhost:8080';
function gotoDemo(page) {
    return __awaiter(this, void 0, void 0, function () {
        var url;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    url = BASE_URL.replace(/\/$/, '') + '/demo.html';
                    return [4 /*yield*/, page.goto(url, { waitUntil: 'load' })];
                case 1:
                    _a.sent();
                    return [4 /*yield*/, page.waitForFunction(function () { return typeof window.JSGantt !== 'undefined'; }, { timeout: 15000 })];
                case 2:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
test_1.test.describe('Issue #349 — vEvents.taskname click handler', function () {
    (0, test_1.test)('taskname click handler fires with task, event and cell arguments', function (_a) {
        var page = _a.page;
        return __awaiter(void 0, void 0, void 0, function () {
            var result, taskCells, count, taskRow, calls, fired, result2, firstTaskCell, callsAfterClick;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, gotoDemo(page)];
                    case 1:
                        _b.sent();
                        return [4 /*yield*/, page.evaluate(function () {
                                var container = document.getElementById('embedded-Gantt');
                                container.innerHTML = '';
                                var calls = [];
                                var g = new window.JSGantt.GanttChart(container, 'week');
                                g.setOptions({
                                    vFormat: 'week',
                                    vShowDeps: 0,
                                    vLang: 'en',
                                    vEvents: {
                                        taskname: function (task, event, cell) {
                                            calls.push({
                                                taskName: task.getName(),
                                                hasEvent: event instanceof MouseEvent,
                                                hasCell: cell instanceof HTMLElement,
                                            });
                                        }
                                    }
                                });
                                var T = window.JSGantt.TaskItem;
                                g.AddTaskItem(new T(1, 'Group A', '2024-01-01', '2024-03-31', 'ggroupblack', '', 0, '', 0, 1, 0, 1, '', '', '', g));
                                g.AddTaskItem(new T(2, 'Task Alpha', '2024-01-01', '2024-01-14', 'gtaskblue', '', 0, 'R1', 0, 0, 1, 1, '', '', '', g));
                                g.AddTaskItem(new T(3, 'Task Beta', '2024-01-15', '2024-02-28', 'gtaskgreen', '', 0, 'R2', 0, 0, 1, 1, '', '', '', g));
                                g.Draw();
                                return { calls: calls, taskCellCount: document.querySelectorAll('.gtaskname').length };
                            })];
                    case 2:
                        result = _b.sent();
                        // Verify task cells rendered
                        (0, test_1.expect)(result.taskCellCount, 'task name cells should be rendered').toBeGreaterThan(0);
                        taskCells = page.locator('.gtaskname');
                        return [4 /*yield*/, taskCells.count()];
                    case 3:
                        count = _b.sent();
                        (0, test_1.expect)(count).toBeGreaterThan(0);
                        taskRow = page.locator('tr.gname .gtaskname').first();
                        return [4 /*yield*/, taskRow.click()];
                    case 4:
                        _b.sent();
                        return [4 /*yield*/, page.evaluate(function () { return window.__issue349Calls; })];
                    case 5:
                        calls = _b.sent();
                        return [4 /*yield*/, page.evaluate(function () {
                                // Re-check by clicking programmatically and capturing via the already-registered handler
                                // Since we registered vEvents.taskname above, clicks on .gtaskname should fire it.
                                // We expose the call log via a global for inspection.
                                return window.__issue349Calls;
                            })];
                    case 6:
                        fired = _b.sent();
                        return [4 /*yield*/, page.evaluate(function () {
                                var container = document.getElementById('embedded-Gantt');
                                container.innerHTML = '';
                                window.__issue349Calls = [];
                                var g = new window.JSGantt.GanttChart(container, 'week');
                                g.setOptions({
                                    vFormat: 'week',
                                    vShowDeps: 0,
                                    vLang: 'en',
                                    vEvents: {
                                        taskname: function (task, event, cell) {
                                            window.__issue349Calls.push({
                                                taskName: task.getName(),
                                                hasEvent: event instanceof MouseEvent,
                                                hasCell: cell instanceof HTMLElement,
                                            });
                                        }
                                    }
                                });
                                var T = window.JSGantt.TaskItem;
                                g.AddTaskItem(new T(1, 'Group A', '2024-01-01', '2024-03-31', 'ggroupblack', '', 0, '', 0, 1, 0, 1, '', '', '', g));
                                g.AddTaskItem(new T(2, 'Task Alpha', '2024-01-01', '2024-01-14', 'gtaskblue', '', 0, 'R1', 0, 0, 1, 1, '', '', '', g));
                                g.AddTaskItem(new T(3, 'Task Beta', '2024-01-15', '2024-02-28', 'gtaskgreen', '', 0, 'R2', 0, 0, 1, 1, '', '', '', g));
                                g.Draw();
                                return { taskCellCount: document.querySelectorAll('tr.gname .gtaskname').length };
                            })];
                    case 7:
                        result2 = _b.sent();
                        (0, test_1.expect)(result2.taskCellCount, 'task rows should be visible').toBeGreaterThan(0);
                        firstTaskCell = page.locator('tr.gname .gtaskname').first();
                        return [4 /*yield*/, firstTaskCell.click()];
                    case 8:
                        _b.sent();
                        return [4 /*yield*/, page.evaluate(function () { return window.__issue349Calls; })];
                    case 9:
                        callsAfterClick = _b.sent();
                        (0, test_1.expect)(callsAfterClick.length, 'taskname handler should fire on click').toBeGreaterThan(0);
                        (0, test_1.expect)(callsAfterClick[0].hasEvent, 'first arg should be a MouseEvent').toBe(true);
                        (0, test_1.expect)(callsAfterClick[0].hasCell, 'second arg should be an HTMLElement').toBe(true);
                        (0, test_1.expect)(typeof callsAfterClick[0].taskName, 'task should have a name').toBe('string');
                        (0, test_1.expect)(callsAfterClick[0].taskName.length, 'task name should be non-empty').toBeGreaterThan(0);
                        return [2 /*return*/];
                }
            });
        });
    });
    (0, test_1.test)('taskname event fires for group task rows', function (_a) {
        var page = _a.page;
        return __awaiter(void 0, void 0, void 0, function () {
            var groupCell, calls;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, gotoDemo(page)];
                    case 1:
                        _b.sent();
                        return [4 /*yield*/, page.evaluate(function () {
                                var container = document.getElementById('embedded-Gantt');
                                container.innerHTML = '';
                                window.__issue349GroupCalls = [];
                                var g = new window.JSGantt.GanttChart(container, 'week');
                                g.setOptions({
                                    vFormat: 'week',
                                    vLang: 'en',
                                    vEvents: {
                                        taskname: function (task, event, cell) {
                                            window.__issue349GroupCalls.push({ taskName: task.getName(), isGroup: task.getGroup() === 1 });
                                        }
                                    }
                                });
                                var T = window.JSGantt.TaskItem;
                                g.AddTaskItem(new T(1, 'Group A', '2024-01-01', '2024-03-31', 'ggroupblack', '', 0, '', 0, 1, 0, 1, '', '', '', g));
                                g.AddTaskItem(new T(2, 'Task Alpha', '2024-01-01', '2024-01-14', 'gtaskblue', '', 0, 'R1', 0, 0, 1, 1, '', '', '', g));
                                g.Draw();
                            })];
                    case 2:
                        _b.sent();
                        groupCell = page.locator('tr.gname .gtaskname div').first();
                        return [4 /*yield*/, groupCell.click()];
                    case 3:
                        _b.sent();
                        return [4 /*yield*/, page.evaluate(function () { return window.__issue349GroupCalls; })];
                    case 4:
                        calls = _b.sent();
                        (0, test_1.expect)(calls.length, 'handler should fire for group row click').toBeGreaterThan(0);
                        (0, test_1.expect)(calls[0].isGroup, 'clicked task should be a group').toBe(true);
                        (0, test_1.expect)(calls[0].taskName).toBe('Group A');
                        return [2 /*return*/];
                }
            });
        });
    });
    (0, test_1.test)('vEvents.taskname can be set via setOptions (not just setEvents directly)', function (_a) {
        var page = _a.page;
        return __awaiter(void 0, void 0, void 0, function () {
            var result, calls;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, gotoDemo(page)];
                    case 1:
                        _b.sent();
                        return [4 /*yield*/, page.evaluate(function () {
                                var container = document.getElementById('embedded-Gantt');
                                container.innerHTML = '';
                                window.__issue349SetOptionsCalls = [];
                                var g = new window.JSGantt.GanttChart(container, 'week');
                                // Use setOptions with vEvents — this is the API shown in issue #349
                                g.setOptions({
                                    vFormat: 'week',
                                    vLang: 'en',
                                    vEvents: {
                                        taskname: function (task) {
                                            window.__issue349SetOptionsCalls.push(task.getName());
                                        }
                                    }
                                });
                                var T = window.JSGantt.TaskItem;
                                g.AddTaskItem(new T(1, 'My Task', '2024-01-01', '2024-01-31', 'gtaskblue', '', 0, '', 0, 0, 0, 1, '', '', '', g));
                                g.Draw();
                                // Verify the events object was set correctly
                                var events = g.getEventsClickCell();
                                return {
                                    hasTasknameHandler: typeof events.taskname === 'function',
                                    taskCellCount: document.querySelectorAll('tr.gname .gtaskname').length,
                                };
                            })];
                    case 2:
                        result = _b.sent();
                        (0, test_1.expect)(result.hasTasknameHandler, 'taskname handler should be registered via setOptions').toBe(true);
                        (0, test_1.expect)(result.taskCellCount).toBeGreaterThan(0);
                        return [4 /*yield*/, page.locator('tr.gname .gtaskname').first().click()];
                    case 3:
                        _b.sent();
                        return [4 /*yield*/, page.evaluate(function () { return window.__issue349SetOptionsCalls; })];
                    case 4:
                        calls = _b.sent();
                        (0, test_1.expect)(calls.length, 'handler should fire').toBeGreaterThan(0);
                        (0, test_1.expect)(calls[0]).toBe('My Task');
                        return [2 /*return*/];
                }
            });
        });
    });
});
