// jsgantt-improved/dist/test/e2e/issue207-bar-axis-alignment.spec.js
"use strict";
/**
 * E2E tests for issue #207: Gantt bars out of sync with timeline axis.
 *
 * The bug: bar left/width positions don't match the header columns
 * for the corresponding start/end dates, causing bars to appear shifted.
 *
 * Strategy:
 *   Inject a deterministic chart (week format) with tasks at known Monday dates.
 *   Read the actual column header labels to locate the correct column for each
 *   task's start date (the chart may add a buffer week, so we do NOT hardcode
 *   column indices — we find them by date text). Then assert each bar's left
 *   edge lands within ALIGN_TOLERANCE_PX of the matching column's left edge.
 *
 *   A systematic drift ≥ colWidth/2 would reproduce issue #207.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var test_1 = require("@playwright/test");
/** Maximum allowed drift between bar left and the corresponding column left (px). */
var ALIGN_TOLERANCE_PX = 4;
/**
 * Navigate to the demo page and wait for the JSGantt global to be available.
 * The demo uses synchronous XHR which fails in headless Chrome, so we skip
 * waiting for the initial chart and inject our own.
 */
function gotoDemo(page) {
    var _a;
    return __awaiter(this, void 0, void 0, function () {
        var baseURL, url;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    baseURL = (_a = process.env.JSGANTT_E2E_BASE_URL) !== null && _a !== void 0 ? _a : 'http://localhost:8080';
                    url = baseURL.replace(/\/$/, '') + '/demo.html';
                    return [4 /*yield*/, page.goto(url, { waitUntil: 'load' })];
                case 1:
                    _b.sent();
                    // Wait for the JSGantt global to be defined (scripts loaded)
                    return [4 /*yield*/, page.waitForFunction(function () { return typeof window.JSGantt !== 'undefined'; }, { timeout: 15000 })];
                case 2:
                    // Wait for the JSGantt global to be defined (scripts loaded)
                    _b.sent();
                    return [2 /*return*/];
            }
        });
    });
}
/** Inject a week-format chart with three 1-week tasks on known Mondays. */
function injectWeekChart(page) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, page.evaluate(function () {
                        var container = document.getElementById('embedded-Gantt');
                        container.innerHTML = '';
                        var g = new window.JSGantt.GanttChart(container, 'week');
                        g.setOptions({ vFormat: 'week', vShowDeps: 0, vLang: 'en', vShowWeekends: 1 });
                        var T = window.JSGantt.TaskItem;
                        // Three 1-week tasks, each starting on a Monday (column boundary)
                        g.AddTaskItem(new T(1, 'Task A', '2024-01-01', '2024-01-07', 'gtaskblue', '', 0, '', 0, 0, 0, 1, '', '', '', g));
                        g.AddTaskItem(new T(2, 'Task B', '2024-01-15', '2024-01-21', 'gtaskgreen', '', 0, '', 0, 0, 0, 1, '', '', '', g));
                        g.AddTaskItem(new T(3, 'Task C', '2024-01-29', '2024-02-04', 'gtaskorange', '', 0, '', 0, 0, 0, 1, '', '', '', g));
                        g.Draw();
                    })];
                case 1:
                    _a.sent();
                    return [4 /*yield*/, page.waitForSelector('#embedded-Gantt .gcharttable', { timeout: 10000 })];
                case 2:
                    _a.sent();
                    return [4 /*yield*/, page.waitForTimeout(400)];
                case 3:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
/** Read bar positions relative to the chart table left edge. */
function getBars(page) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, page.evaluate(function () {
                    var chartRect = document.querySelector('.gcharttable').getBoundingClientRect();
                    return Array.from(document.querySelectorAll('[id*="bardiv_"]'))
                        .filter(function (el) { return !el.classList.contains('gplan'); })
                        .map(function (el) { return ({
                        id: el.id,
                        left: Math.round((el.getBoundingClientRect().left - chartRect.left) * 10) / 10,
                    }); });
                })];
        });
    });
}
/** Read header column positions and date labels. */
function getCols(page) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, page.evaluate(function () {
                    var headerTable = document.querySelector('.gcharttableh');
                    var headerRect = headerTable.getBoundingClientRect();
                    var rows = headerTable.querySelectorAll('tr');
                    var lastRow = rows[rows.length - 1]; // finest granularity row
                    return Array.from(lastRow.querySelectorAll('td')).map(function (td, i) {
                        var r = td.getBoundingClientRect();
                        return {
                            index: i,
                            text: td.innerText.trim().slice(0, 10),
                            left: Math.round((r.left - headerRect.left) * 10) / 10,
                            width: Math.round(r.width * 10) / 10,
                        };
                    });
                })];
        });
    });
}
/**
 * Find the column whose label starts with the given dd/mm string.
 * Week headers are rendered as "DD/MM" in the default date display format.
 */
function findColByDate(cols, dd, mm) {
    var target = "".concat(String(dd).padStart(2, '0'), "/").concat(String(mm).padStart(2, '0'));
    return cols.find(function (c) { return c.text.startsWith(target); });
}
test_1.test.describe('Issue #207 — bar alignment with timeline axis (week format)', function () {
    (0, test_1.test)('Task A (01/01): bar left aligns with the 01/01 column', function (_a) {
        var page = _a.page;
        return __awaiter(void 0, void 0, void 0, function () {
            var bars, cols, col, bar, drift;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, gotoDemo(page)];
                    case 1:
                        _b.sent();
                        return [4 /*yield*/, injectWeekChart(page)];
                    case 2:
                        _b.sent();
                        return [4 /*yield*/, getBars(page)];
                    case 3:
                        bars = _b.sent();
                        return [4 /*yield*/, getCols(page)];
                    case 4:
                        cols = _b.sent();
                        (0, test_1.expect)(bars.length, 'should find rendered bars').toBeGreaterThanOrEqual(3);
                        col = findColByDate(cols, 1, 1);
                        (0, test_1.expect)(col, 'should find a column labelled 01/01').toBeDefined();
                        bar = bars[0];
                        drift = Math.abs(bar.left - col.left);
                        (0, test_1.expect)(drift, "Task A bar left (".concat(bar.left, "px) should align with col \"01/01\" left (").concat(col.left, "px), drift=").concat(drift.toFixed(1), "px")).toBeLessThanOrEqual(ALIGN_TOLERANCE_PX);
                        return [2 /*return*/];
                }
            });
        });
    });
    (0, test_1.test)('Task B (15/01): bar left aligns with the 15/01 column', function (_a) {
        var page = _a.page;
        return __awaiter(void 0, void 0, void 0, function () {
            var bars, cols, col, bar, drift;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, gotoDemo(page)];
                    case 1:
                        _b.sent();
                        return [4 /*yield*/, injectWeekChart(page)];
                    case 2:
                        _b.sent();
                        return [4 /*yield*/, getBars(page)];
                    case 3:
                        bars = _b.sent();
                        return [4 /*yield*/, getCols(page)];
                    case 4:
                        cols = _b.sent();
                        col = findColByDate(cols, 15, 1);
                        (0, test_1.expect)(col, 'should find a column labelled 15/01').toBeDefined();
                        bar = bars[1];
                        drift = Math.abs(bar.left - col.left);
                        (0, test_1.expect)(drift, "Task B bar left (".concat(bar.left, "px) should align with col \"15/01\" left (").concat(col.left, "px), drift=").concat(drift.toFixed(1), "px")).toBeLessThanOrEqual(ALIGN_TOLERANCE_PX);
                        return [2 /*return*/];
                }
            });
        });
    });
    (0, test_1.test)('Task C (29/01): bar left aligns with the 29/01 column', function (_a) {
        var page = _a.page;
        return __awaiter(void 0, void 0, void 0, function () {
            var bars, cols, col, bar, drift;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, gotoDemo(page)];
                    case 1:
                        _b.sent();
                        return [4 /*yield*/, injectWeekChart(page)];
                    case 2:
                        _b.sent();
                        return [4 /*yield*/, getBars(page)];
                    case 3:
                        bars = _b.sent();
                        return [4 /*yield*/, getCols(page)];
                    case 4:
                        cols = _b.sent();
                        col = findColByDate(cols, 29, 1);
                        (0, test_1.expect)(col, 'should find a column labelled 29/01').toBeDefined();
                        bar = bars[2];
                        drift = Math.abs(bar.left - col.left);
                        (0, test_1.expect)(drift, "Task C bar left (".concat(bar.left, "px) should align with col \"29/01\" left (").concat(col.left, "px), drift=").concat(drift.toFixed(1), "px")).toBeLessThanOrEqual(ALIGN_TOLERANCE_PX);
                        return [2 /*return*/];
                }
            });
        });
    });
    (0, test_1.test)('each bar left is exactly on a column boundary (not between columns)', function (_a) {
        var page = _a.page;
        return __awaiter(void 0, void 0, void 0, function () {
            var bars, cols, _loop_1, _i, bars_1, bar;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, gotoDemo(page)];
                    case 1:
                        _b.sent();
                        return [4 /*yield*/, injectWeekChart(page)];
                    case 2:
                        _b.sent();
                        return [4 /*yield*/, getBars(page)];
                    case 3:
                        bars = _b.sent();
                        return [4 /*yield*/, getCols(page)];
                    case 4:
                        cols = _b.sent();
                        (0, test_1.expect)(cols.length).toBeGreaterThan(0);
                        _loop_1 = function (bar) {
                            var minDrift = Math.min.apply(Math, cols.map(function (c) { return Math.abs(bar.left - c.left); }));
                            (0, test_1.expect)(minDrift, "bar ".concat(bar.id, " at ").concat(bar.left, "px should sit on a column boundary; nearest column drift=").concat(minDrift.toFixed(1), "px")).toBeLessThanOrEqual(ALIGN_TOLERANCE_PX);
                        };
                        for (_i = 0, bars_1 = bars; _i < bars_1.length; _i++) {
                            bar = bars_1[_i];
                            _loop_1(bar);
                        }
                        return [2 /*return*/];
                }
            });
        });
    });
    (0, test_1.test)('month format (react-jsgantt scenario): bar aligns with 01/01 column', function (_a) {
        var page = _a.page;
        return __awaiter(void 0, void 0, void 0, function () {
            var bars, cols, _loop_2, _i, bars_2, bar;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, gotoDemo(page)];
                    case 1:
                        _b.sent();
                        // Reproduce the react-jsgantt scenario: month-format tasks at exact month boundaries
                        return [4 /*yield*/, page.evaluate(function () {
                                var container = document.getElementById('embedded-Gantt');
                                container.innerHTML = '';
                                var g = new window.JSGantt.GanttChart(container, 'month');
                                g.setOptions({ vFormat: 'month', vShowDeps: 0, vLang: 'en' });
                                var T = window.JSGantt.TaskItem;
                                g.AddTaskItem(new T(1, 'Phase 1', '2024-01-01', '2024-01-31', 'gtaskblue', '', 0, '', 0, 0, 0, 1, '', '', '', g));
                                g.AddTaskItem(new T(2, 'Phase 2', '2024-03-01', '2024-03-31', 'gtaskgreen', '', 0, '', 0, 0, 0, 1, '', '', '', g));
                                g.Draw();
                            })];
                    case 2:
                        // Reproduce the react-jsgantt scenario: month-format tasks at exact month boundaries
                        _b.sent();
                        return [4 /*yield*/, page.waitForSelector('#embedded-Gantt .gcharttable', { timeout: 10000 })];
                    case 3:
                        _b.sent();
                        return [4 /*yield*/, page.waitForTimeout(400)];
                    case 4:
                        _b.sent();
                        return [4 /*yield*/, getBars(page)];
                    case 5:
                        bars = _b.sent();
                        return [4 /*yield*/, getCols(page)];
                    case 6:
                        cols = _b.sent();
                        (0, test_1.expect)(bars.length).toBeGreaterThanOrEqual(2);
                        _loop_2 = function (bar) {
                            var minDrift = Math.min.apply(Math, cols.map(function (c) { return Math.abs(bar.left - c.left); }));
                            (0, test_1.expect)(minDrift, "bar ".concat(bar.id, " at ").concat(bar.left, "px should sit on a column boundary; nearest drift=").concat(minDrift.toFixed(1), "px")).toBeLessThanOrEqual(ALIGN_TOLERANCE_PX);
                        };
                        // Each bar should sit on a column boundary
                        for (_i = 0, bars_2 = bars; _i < bars_2.length; _i++) {
                            bar = bars_2[_i];
                            _loop_2(bar);
                        }
                        return [2 /*return*/];
                }
            });
        });
    });
});
