// jsgantt-improved/dist-e2e/issue207-bar-axis-alignment.spec.js
"use strict";
/**
 * E2E tests for issue #207: Gantt bars out of sync with timeline axis.
 *
 * The bug: bar left/width positions don't match the header columns
 * for the corresponding start/end dates, causing bars to appear shifted.
 *
 * Strategy:
 *   Inject a deterministic chart (week format) with tasks at known Monday dates.
 *   Read the actual column header labels to locate the correct column for each
 *   task's start date (the chart may add a buffer week, so we do NOT hardcode
 *   column indices — we find them by date text). Then assert each bar's left
 *   edge lands within ALIGN_TOLERANCE_PX of the matching column's left edge.
 *
 *   A systematic drift ≥ colWidth/2 would reproduce issue #207.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const test_1 = require("@playwright/test");
/** Maximum allowed drift between bar left and the corresponding column left (px). */
const ALIGN_TOLERANCE_PX = 4;
/**
 * Navigate to the demo page and wait for the JSGantt global to be available.
 * The demo uses synchronous XHR which fails in headless Chrome, so we skip
 * waiting for the initial chart and inject our own.
 */
async function gotoDemo(page) {
    var _a;
    // Use an absolute URL so Playwright doesn't reject "demo.html" as invalid
    const baseURL = (_a = process.env.JSGANTT_E2E_BASE_URL) !== null && _a !== void 0 ? _a : 'http://localhost:8080';
    const url = baseURL.replace(/\/$/, '') + '/demo.html';
    await page.goto(url, { waitUntil: 'load' });
    // Wait for the JSGantt global to be defined (scripts loaded)
    await page.waitForFunction(() => typeof window.JSGantt !== 'undefined', { timeout: 15000 });
}
/** Inject a week-format chart with three 1-week tasks on known Mondays. */
async function injectWeekChart(page) {
    await page.evaluate(() => {
        const container = document.getElementById('embedded-Gantt');
        container.innerHTML = '';
        const g = new window.JSGantt.GanttChart(container, 'week');
        g.setOptions({ vFormat: 'week', vShowDeps: 0, vLang: 'en', vShowWeekends: 1 });
        const T = window.JSGantt.TaskItem;
        // Three 1-week tasks, each starting on a Monday (column boundary)
        g.AddTaskItem(new T(1, 'Task A', '2024-01-01', '2024-01-07', 'gtaskblue', '', 0, '', 0, 0, 0, 1, '', '', '', g));
        g.AddTaskItem(new T(2, 'Task B', '2024-01-15', '2024-01-21', 'gtaskgreen', '', 0, '', 0, 0, 0, 1, '', '', '', g));
        g.AddTaskItem(new T(3, 'Task C', '2024-01-29', '2024-02-04', 'gtaskorange', '', 0, '', 0, 0, 0, 1, '', '', '', g));
        g.Draw();
    });
    await page.waitForSelector('#embedded-Gantt .gcharttable', { timeout: 10000 });
    await page.waitForTimeout(400);
}
/** Read bar positions relative to the chart table left edge. */
async function getBars(page) {
    return page.evaluate(() => {
        const chartRect = document.querySelector('.gcharttable').getBoundingClientRect();
        return Array.from(document.querySelectorAll('[id*="bardiv_"]'))
            .filter(el => !el.classList.contains('gplan'))
            .map(el => ({
            id: el.id,
            left: Math.round((el.getBoundingClientRect().left - chartRect.left) * 10) / 10,
        }));
    });
}
/** Read header column positions and date labels. */
async function getCols(page) {
    return page.evaluate(() => {
        const headerTable = document.querySelector('.gcharttableh');
        const headerRect = headerTable.getBoundingClientRect();
        const rows = headerTable.querySelectorAll('tr');
        const lastRow = rows[rows.length - 1]; // finest granularity row
        return Array.from(lastRow.querySelectorAll('td')).map((td, i) => {
            const r = td.getBoundingClientRect();
            return {
                index: i,
                text: td.innerText.trim().slice(0, 10),
                left: Math.round((r.left - headerRect.left) * 10) / 10,
                width: Math.round(r.width * 10) / 10,
            };
        });
    });
}
/**
 * Find the column whose label starts with the given dd/mm string.
 * Week headers are rendered as "DD/MM" in the default date display format.
 */
function findColByDate(cols, dd, mm) {
    const target = `${String(dd).padStart(2, '0')}/${String(mm).padStart(2, '0')}`;
    return cols.find(c => c.text.startsWith(target));
}
test_1.test.describe('Issue #207 — bar alignment with timeline axis (week format)', () => {
    (0, test_1.test)('Task A (01/01): bar left aligns with the 01/01 column', async ({ page }) => {
        await gotoDemo(page);
        await injectWeekChart(page);
        const bars = await getBars(page);
        const cols = await getCols(page);
        (0, test_1.expect)(bars.length, 'should find rendered bars').toBeGreaterThanOrEqual(3);
        const col = findColByDate(cols, 1, 1);
        (0, test_1.expect)(col, 'should find a column labelled 01/01').toBeDefined();
        const bar = bars[0];
        const drift = Math.abs(bar.left - col.left);
        (0, test_1.expect)(drift, `Task A bar left (${bar.left}px) should align with col "01/01" left (${col.left}px), drift=${drift.toFixed(1)}px`).toBeLessThanOrEqual(ALIGN_TOLERANCE_PX);
    });
    (0, test_1.test)('Task B (15/01): bar left aligns with the 15/01 column', async ({ page }) => {
        await gotoDemo(page);
        await injectWeekChart(page);
        const bars = await getBars(page);
        const cols = await getCols(page);
        const col = findColByDate(cols, 15, 1);
        (0, test_1.expect)(col, 'should find a column labelled 15/01').toBeDefined();
        const bar = bars[1];
        const drift = Math.abs(bar.left - col.left);
        (0, test_1.expect)(drift, `Task B bar left (${bar.left}px) should align with col "15/01" left (${col.left}px), drift=${drift.toFixed(1)}px`).toBeLessThanOrEqual(ALIGN_TOLERANCE_PX);
    });
    (0, test_1.test)('Task C (29/01): bar left aligns with the 29/01 column', async ({ page }) => {
        await gotoDemo(page);
        await injectWeekChart(page);
        const bars = await getBars(page);
        const cols = await getCols(page);
        const col = findColByDate(cols, 29, 1);
        (0, test_1.expect)(col, 'should find a column labelled 29/01').toBeDefined();
        const bar = bars[2];
        const drift = Math.abs(bar.left - col.left);
        (0, test_1.expect)(drift, `Task C bar left (${bar.left}px) should align with col "29/01" left (${col.left}px), drift=${drift.toFixed(1)}px`).toBeLessThanOrEqual(ALIGN_TOLERANCE_PX);
    });
    (0, test_1.test)('each bar left is exactly on a column boundary (not between columns)', async ({ page }) => {
        await gotoDemo(page);
        await injectWeekChart(page);
        const bars = await getBars(page);
        const cols = await getCols(page);
        (0, test_1.expect)(cols.length).toBeGreaterThan(0);
        for (const bar of bars) {
            const minDrift = Math.min(...cols.map(c => Math.abs(bar.left - c.left)));
            (0, test_1.expect)(minDrift, `bar ${bar.id} at ${bar.left}px should sit on a column boundary; nearest column drift=${minDrift.toFixed(1)}px`).toBeLessThanOrEqual(ALIGN_TOLERANCE_PX);
        }
    });
    (0, test_1.test)('month format (react-jsgantt scenario): bar aligns with 01/01 column', async ({ page }) => {
        await gotoDemo(page);
        // Reproduce the react-jsgantt scenario: month-format tasks at exact month boundaries
        await page.evaluate(() => {
            const container = document.getElementById('embedded-Gantt');
            container.innerHTML = '';
            const g = new window.JSGantt.GanttChart(container, 'month');
            g.setOptions({ vFormat: 'month', vShowDeps: 0, vLang: 'en' });
            const T = window.JSGantt.TaskItem;
            g.AddTaskItem(new T(1, 'Phase 1', '2024-01-01', '2024-01-31', 'gtaskblue', '', 0, '', 0, 0, 0, 1, '', '', '', g));
            g.AddTaskItem(new T(2, 'Phase 2', '2024-03-01', '2024-03-31', 'gtaskgreen', '', 0, '', 0, 0, 0, 1, '', '', '', g));
            g.Draw();
        });
        await page.waitForSelector('#embedded-Gantt .gcharttable', { timeout: 10000 });
        await page.waitForTimeout(400);
        const bars = await getBars(page);
        const cols = await getCols(page);
        (0, test_1.expect)(bars.length).toBeGreaterThanOrEqual(2);
        // Each bar should sit on a column boundary
        for (const bar of bars) {
            const minDrift = Math.min(...cols.map(c => Math.abs(bar.left - c.left)));
            (0, test_1.expect)(minDrift, `bar ${bar.id} at ${bar.left}px should sit on a column boundary; nearest drift=${minDrift.toFixed(1)}px`).toBeLessThanOrEqual(ALIGN_TOLERANCE_PX);
        }
    });
});
