// jsgantt-improved/dist-e2e/issue-setMinDate-setMaxDate.spec.js
"use strict";
/**
 * E2E tests for setMinDate / setMaxDate via setOptions.
 *
 * Bug: passing vMinDate or vMaxDate to setOptions threw
 *   "TypeError: this.setMinDate is not a function"
 * because no setter existed on GanttChart for those options.
 *
 * Fix: added setMinDate / setMaxDate setters that store the override date string,
 * and updated Draw() to use them (falling back to auto-computed dates when empty).
 */
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
const test_1 = require("@playwright/test");
const BASE_URL = (_a = process.env.JSGANTT_E2E_BASE_URL) !== null && _a !== void 0 ? _a : 'http://localhost:8080';
async function gotoDemo(page) {
    const url = BASE_URL.replace(/\/$/, '') + '/demo.html';
    await page.goto(url, { waitUntil: 'load' });
    await page.waitForFunction(() => typeof window.JSGantt !== 'undefined', { timeout: 15000 });
}
test_1.test.describe('setMinDate / setMaxDate via setOptions', () => {
    test_1.test('setOptions with vMinDate and vMaxDate does not throw', async ({ page }) => {
        await gotoDemo(page);
        const result = await page.evaluate(() => {
            const container = document.getElementById('embedded-Gantt');
            container.innerHTML = '';
            let error = null;
            try {
                const g = new window.JSGantt.GanttChart(container, 'week');
                g.setOptions({
                    vDateInputFormat: 'yyyy-mm-dd',
                    vMinDate: '2024-01-01',
                    vMaxDate: '2024-06-30',
                    vFormat: 'week',
                });
                const T = window.JSGantt.TaskItem;
                g.AddTaskItem(new T(1, 'Task A', '2024-02-01', '2024-04-30', 'gtaskblue', '', 0, '', 0, 0, 0, 1, '', '', '', g));
                g.Draw();
            } catch (e) {
                error = e.message;
            }
            return { error };
        });
        test_1.expect(result.error, 'setOptions with vMinDate/vMaxDate should not throw').toBeNull();
    });

    test_1.test('vMinDate overrides the chart start boundary', async ({ page }) => {
        await gotoDemo(page);
        const result = await page.evaluate(() => {
            const container = document.getElementById('embedded-Gantt');
            container.innerHTML = '';
            const g = new window.JSGantt.GanttChart(container, 'week');
            g.setOptions({
                vDateInputFormat: 'yyyy-mm-dd',
                vMinDate: '2024-01-01',
                vMaxDate: '2024-12-31',
                vFormat: 'week',
                vShowDeps: 0,
            });
            const T = window.JSGantt.TaskItem;
            // Task starts in March — with vMinDate=Jan 1, the chart should start before March
            g.AddTaskItem(new T(1, 'Task A', '2024-03-01', '2024-05-31', 'gtaskblue', '', 0, '', 0, 0, 0, 1, '', '', '', g));
            g.Draw();
            // Check that chart head table was rendered inside our container
            const chartHead = container.querySelector('.gcharthead');
            const chartTable = container.querySelector('.gcharttable');
            const allText = container.innerHTML;
            return {
                hasChartHead: !!chartHead,
                hasChartTable: !!chartTable,
                contains2024: allText.includes('2024'),
            };
        });
        test_1.expect(result.hasChartTable, 'chart table should be rendered').toBe(true);
        test_1.expect(result.contains2024, 'chart should show year 2024 from vMinDate/vMaxDate range').toBe(true);
    });

    test_1.test('getMinDate and getMaxDate return the set values', async ({ page }) => {
        await gotoDemo(page);
        const result = await page.evaluate(() => {
            const container = document.getElementById('embedded-Gantt');
            container.innerHTML = '';
            const g = new window.JSGantt.GanttChart(container, 'week');
            g.setOptions({
                vMinDate: '2024-01-01',
                vMaxDate: '2024-12-31',
            });
            return {
                minDate: g.getMinDate(),
                maxDate: g.getMaxDate(),
            };
        });
        test_1.expect(result.minDate).toBe('2024-01-01');
        test_1.expect(result.maxDate).toBe('2024-12-31');
    });
});
