// jsgantt-improved/dist-e2e/issue349-taskname-click-event.spec.js
"use strict";
/**
 * E2E tests for issue #349: Click event on task name via vEvents.taskname.
 *
 * The issue reports that users want to handle clicks on task name cells.
 * The vEvents.taskname callback should fire with (task, event, cell) args
 * when a task name cell is clicked.
 */
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
const test_1 = require("@playwright/test");
const BASE_URL = (_a = process.env.JSGANTT_E2E_BASE_URL) !== null && _a !== void 0 ? _a : 'http://localhost:8080';
async function gotoDemo(page) {
    const url = BASE_URL.replace(/\/$/, '') + '/demo.html';
    await page.goto(url, { waitUntil: 'load' });
    await page.waitForFunction(() => typeof window.JSGantt !== 'undefined', { timeout: 15000 });
}
test_1.test.describe('Issue #349 — vEvents.taskname click handler', () => {
    (0, test_1.test)('taskname click handler fires with task, event and cell arguments', async ({ page }) => {
        await gotoDemo(page);
        const result = await page.evaluate(() => {
            const container = document.getElementById('embedded-Gantt');
            container.innerHTML = '';
            const calls = [];
            const g = new window.JSGantt.GanttChart(container, 'week');
            g.setOptions({
                vFormat: 'week',
                vShowDeps: 0,
                vLang: 'en',
                vEvents: {
                    taskname: function (task, event, cell) {
                        calls.push({
                            taskName: task.getName(),
                            hasEvent: event instanceof MouseEvent,
                            hasCell: cell instanceof HTMLElement,
                        });
                    }
                }
            });
            const T = window.JSGantt.TaskItem;
            g.AddTaskItem(new T(1, 'Group A', '2024-01-01', '2024-03-31', 'ggroupblack', '', 0, '', 0, 1, 0, 1, '', '', '', g));
            g.AddTaskItem(new T(2, 'Task Alpha', '2024-01-01', '2024-01-14', 'gtaskblue', '', 0, 'R1', 0, 0, 1, 1, '', '', '', g));
            g.AddTaskItem(new T(3, 'Task Beta', '2024-01-15', '2024-02-28', 'gtaskgreen', '', 0, 'R2', 0, 0, 1, 1, '', '', '', g));
            g.Draw();
            return { calls, taskCellCount: document.querySelectorAll('.gtaskname').length };
        });
        // Verify task cells rendered
        (0, test_1.expect)(result.taskCellCount, 'task name cells should be rendered').toBeGreaterThan(0);
        // Click on task name cells
        const taskCells = page.locator('.gtaskname');
        const count = await taskCells.count();
        (0, test_1.expect)(count).toBeGreaterThan(0);
        // Click the first non-header task cell (skip the header cell)
        // Task list rows have .gname class on their tr; click cell div inside
        const taskRow = page.locator('tr.gname .gtaskname').first();
        await taskRow.click();
        // Read captured calls from page
        const calls = await page.evaluate(() => window.__issue349Calls);
        // The calls array was captured inside evaluate scope, re-check via a different approach
        const fired = await page.evaluate(() => {
            // Re-check by clicking programmatically and capturing via the already-registered handler
            // Since we registered vEvents.taskname above, clicks on .gtaskname should fire it.
            // We expose the call log via a global for inspection.
            return window.__issue349Calls;
        });
        // Since our calls array is inside the evaluate closure, we need to re-run with a global
        // Let's re-run with a global-exposed approach
        const result2 = await page.evaluate(() => {
            const container = document.getElementById('embedded-Gantt');
            container.innerHTML = '';
            window.__issue349Calls = [];
            const g = new window.JSGantt.GanttChart(container, 'week');
            g.setOptions({
                vFormat: 'week',
                vShowDeps: 0,
                vLang: 'en',
                vEvents: {
                    taskname: function (task, event, cell) {
                        window.__issue349Calls.push({
                            taskName: task.getName(),
                            hasEvent: event instanceof MouseEvent,
                            hasCell: cell instanceof HTMLElement,
                        });
                    }
                }
            });
            const T = window.JSGantt.TaskItem;
            g.AddTaskItem(new T(1, 'Group A', '2024-01-01', '2024-03-31', 'ggroupblack', '', 0, '', 0, 1, 0, 1, '', '', '', g));
            g.AddTaskItem(new T(2, 'Task Alpha', '2024-01-01', '2024-01-14', 'gtaskblue', '', 0, 'R1', 0, 0, 1, 1, '', '', '', g));
            g.AddTaskItem(new T(3, 'Task Beta', '2024-01-15', '2024-02-28', 'gtaskgreen', '', 0, 'R2', 0, 0, 1, 1, '', '', '', g));
            g.Draw();
            return { taskCellCount: document.querySelectorAll('tr.gname .gtaskname').length };
        });
        (0, test_1.expect)(result2.taskCellCount, 'task rows should be visible').toBeGreaterThan(0);
        // Click the first task cell (non-group row)
        const firstTaskCell = page.locator('tr.gname .gtaskname').first();
        await firstTaskCell.click();
        const callsAfterClick = await page.evaluate(() => window.__issue349Calls);
        (0, test_1.expect)(callsAfterClick.length, 'taskname handler should fire on click').toBeGreaterThan(0);
        (0, test_1.expect)(callsAfterClick[0].hasEvent, 'first arg should be a MouseEvent').toBe(true);
        (0, test_1.expect)(callsAfterClick[0].hasCell, 'second arg should be an HTMLElement').toBe(true);
        (0, test_1.expect)(typeof callsAfterClick[0].taskName, 'task should have a name').toBe('string');
        (0, test_1.expect)(callsAfterClick[0].taskName.length, 'task name should be non-empty').toBeGreaterThan(0);
    });
    (0, test_1.test)('taskname event fires for group task rows', async ({ page }) => {
        await gotoDemo(page);
        await page.evaluate(() => {
            const container = document.getElementById('embedded-Gantt');
            container.innerHTML = '';
            window.__issue349GroupCalls = [];
            const g = new window.JSGantt.GanttChart(container, 'week');
            g.setOptions({
                vFormat: 'week',
                vLang: 'en',
                vEvents: {
                    taskname: function (task, event, cell) {
                        window.__issue349GroupCalls.push({ taskName: task.getName(), isGroup: task.getGroup() === 1 });
                    }
                }
            });
            const T = window.JSGantt.TaskItem;
            g.AddTaskItem(new T(1, 'Group A', '2024-01-01', '2024-03-31', 'ggroupblack', '', 0, '', 0, 1, 0, 1, '', '', '', g));
            g.AddTaskItem(new T(2, 'Task Alpha', '2024-01-01', '2024-01-14', 'gtaskblue', '', 0, 'R1', 0, 0, 1, 1, '', '', '', g));
            g.Draw();
        });
        // Click the group row's taskname cell
        // Group rows have a div with a span for the collapse icon inside .gtaskname
        const groupCell = page.locator('tr.gname .gtaskname div').first();
        await groupCell.click();
        const calls = await page.evaluate(() => window.__issue349GroupCalls);
        (0, test_1.expect)(calls.length, 'handler should fire for group row click').toBeGreaterThan(0);
        (0, test_1.expect)(calls[0].isGroup, 'clicked task should be a group').toBe(true);
        (0, test_1.expect)(calls[0].taskName).toBe('Group A');
    });
    (0, test_1.test)('vEvents.taskname can be set via setOptions (not just setEvents directly)', async ({ page }) => {
        await gotoDemo(page);
        const result = await page.evaluate(() => {
            const container = document.getElementById('embedded-Gantt');
            container.innerHTML = '';
            window.__issue349SetOptionsCalls = [];
            const g = new window.JSGantt.GanttChart(container, 'week');
            // Use setOptions with vEvents — this is the API shown in issue #349
            g.setOptions({
                vFormat: 'week',
                vLang: 'en',
                vEvents: {
                    taskname: function (task) {
                        window.__issue349SetOptionsCalls.push(task.getName());
                    }
                }
            });
            const T = window.JSGantt.TaskItem;
            g.AddTaskItem(new T(1, 'My Task', '2024-01-01', '2024-01-31', 'gtaskblue', '', 0, '', 0, 0, 0, 1, '', '', '', g));
            g.Draw();
            // Verify the events object was set correctly
            const events = g.getEventsClickCell();
            return {
                hasTasknameHandler: typeof events.taskname === 'function',
                taskCellCount: document.querySelectorAll('tr.gname .gtaskname').length,
            };
        });
        (0, test_1.expect)(result.hasTasknameHandler, 'taskname handler should be registered via setOptions').toBe(true);
        (0, test_1.expect)(result.taskCellCount).toBeGreaterThan(0);
        await page.locator('tr.gname .gtaskname').first().click();
        const calls = await page.evaluate(() => window.__issue349SetOptionsCalls);
        (0, test_1.expect)(calls.length, 'handler should fire').toBeGreaterThan(0);
        (0, test_1.expect)(calls[0]).toBe('My Task');
    });
});
