// jsgantt-improved/dist/test/unit/gantt-chart-constructor.spec.js
"use strict";
/**
 * Unit tests for GanttChart constructor — jQuery-element normalisation.
 *
 * Issue #344: passing a jQuery-wrapped element threw "hasChildNodes is not a
 * function" because pDiv[0] (the real Element) was never unwrapped.
 *
 * PR #393 fixes this with the guard:
 *   this.vDiv = (pDiv && pDiv[0] instanceof Element) ? pDiv[0] : pDiv;
 */
Object.defineProperty(exports, "__esModule", { value: true });
require("./helpers"); // DOM shim — must be first
var chai_1 = require("chai");
// GanttChart is exported as a plain constructor function.
var draw_1 = require("../../src/draw");
// ─── Minimal fake DOM element ─────────────────────────────────────────────────
// The shim in helpers.ts installs document.createElement / createTextNode so
// that TaskItem works.  GanttChart only stores pDiv in this.vDiv, so a plain
// object that satisfies `instanceof Element` suffices.
function makeFakeElement() {
    // jsdom is not available; create a minimal stand-in that IS an Element by
    // delegating through the global.Element constructor set up in the shim.
    // However, `instanceof Element` requires the *global* Element constructor.
    // We set one up below so tests are self-contained.
    return new global.Element();
}
// Install a minimal Element constructor in the global scope so that
// `pDiv[0] instanceof Element` works inside GanttChart.
before(function () {
    if (!global.Element) {
        global.Element = /** @class */ (function () {
            function Element() {
                this.nodeName = 'DIV';
                this.className = '';
                this.childNodes = [];
            }
            Element.prototype.hasChildNodes = function () { return this.childNodes.length > 0; };
            return Element;
        }());
    }
});
// ─── Tests ────────────────────────────────────────────────────────────────────
describe('GanttChart constructor — pDiv normalisation (issue #344)', function () {
    it('stores a plain DOM element unchanged', function () {
        var el = makeFakeElement();
        var chart = new draw_1.GanttChart(el, 'week');
        (0, chai_1.expect)(chart.vDiv).to.equal(el);
    });
    it('unwraps a jQuery-like object (pDiv[0] instanceof Element)', function () {
        var el = makeFakeElement();
        // Simulate a jQuery wrapper: array-like with [0] pointing to the element.
        var jqueryLike = { 0: el, length: 1 };
        var chart = new draw_1.GanttChart(jqueryLike, 'week');
        (0, chai_1.expect)(chart.vDiv).to.equal(el, 'vDiv should be the unwrapped DOM element, not the jQuery wrapper');
    });
    it('does NOT unwrap when pDiv[0] is not an Element', function () {
        // e.g. someone passes an arbitrary array-like whose [0] is a string
        var notAnElement = { 0: 'string-value', length: 1 };
        var chart = new draw_1.GanttChart(notAnElement, 'week');
        (0, chai_1.expect)(chart.vDiv).to.equal(notAnElement, 'vDiv should remain unchanged when [0] is not an Element');
    });
    it('handles null pDiv without throwing', function () {
        (0, chai_1.expect)(function () { return new draw_1.GanttChart(null, 'week'); }).to.not.throw();
        var chart = new draw_1.GanttChart(null, 'week');
        (0, chai_1.expect)(chart.vDiv).to.equal(null);
    });
    it('handles undefined pDiv without throwing', function () {
        (0, chai_1.expect)(function () { return new draw_1.GanttChart(undefined, 'week'); }).to.not.throw();
        var chart = new draw_1.GanttChart(undefined, 'week');
        // undefined is falsy → guard short-circuits → vDiv stays undefined
        (0, chai_1.expect)(chart.vDiv).to.equal(undefined);
    });
    it('stores the format string in vFormat', function () {
        var el = makeFakeElement();
        var chart = new draw_1.GanttChart(el, 'month');
        (0, chai_1.expect)(chart.vFormat).to.equal('month');
    });
    it('stores the correct format when a jQuery wrapper is passed', function () {
        var el = makeFakeElement();
        var jqueryLike = { 0: el, length: 1 };
        var chart = new draw_1.GanttChart(jqueryLike, 'day');
        (0, chai_1.expect)(chart.vFormat).to.equal('day');
    });
});
